### ISSUE TITLE
Kubelet stops rotating service account tokens when pod is terminating, breaking preStop hooks

### ISSUE THREAD

**Issue by PaulFurtado** [kubernetes/kubernetes#116481]
### What happened?

We have pods that use the kube API during their preStop hooks and we use 12h service account token expirations. Today we had a pod where the token was approaching expiration at the time the pod was deleted. 4h into the preStop hook, the token expired and the preStop hook started failing.

### What did you expect to happen?

Service account token rotation should continue working until pods have fully terminated.

### How can we reproduce it (as minimally and precisely as possible)?

Run a pod with a 10-minute service account token expiration and long terminationGracePeriodSeconds and a preStop hook that does `sleep infinity`. Exec into the pod after 10 minutes have elapsed and check the token, it will be expired.

### Kubernetes version

<details>

```console
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.16", GitCommit:"60e5135f758b6e43d0523b3277e8d34b4ab3801f", GitTreeState:"clean", BuildDate:"2023-01-18T16:01:10Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.16", GitCommit:"60e5135f758b6e43d0523b3277e8d34b4ab3801f", GitTreeState:"archive", BuildDate:"2023-03-01T14:28:40Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
AWS, but N/A for this issue
</details>

### OS version

<details>
CentOS 8 stream, but N/A for this issue

</details>

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

Occurs with both cri-o and cri-dockerd

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

**Comment by smarterclayton** [comment]
/cc

This may be related to #115342, but if not I want to see why this still happens.

**Comment by liggitt** [comment]
Do all volume refreshes stop for terminating pods?

**Comment by chrisayoub** [comment]
@liggitt It seems so. 

In kubelet's `syncPod` (which based on verbose logs, seems to execute regularly for running pods), it specifically stops reconciling volumes for pods once the pod enters the terminating state:
- https://github.com/kubernetes/kubernetes/blob/ead7d66ee12656cfb7c633dd42a87f8d9cfaa469/pkg/kubelet/kubelet.go#L1820
- https://github.com/kubernetes/kubernetes/blob/ead7d66ee12656cfb7c633dd42a87f8d9cfaa469/pkg/kubelet/volumemanager/volume_manager.go#L405
- Note the comment here: https://github.com/kubernetes/kubernetes/blob/ead7d66ee12656cfb7c633dd42a87f8d9cfaa469/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L330

**Comment by liggitt** [comment]
That doesn't seem correct... if a container hasn't yet started or is already terminated, it makes sense not to try to manage volumes for it. If it is running, things like configmaps/secrets/tokens still being kept updated seems correct.

**Comment by PaulFurtado** [comment]
> That doesn't seem correct... if a container hasn't yet started or is already terminated, it makes sense not to try to manage volumes for it. If it is running, things like configmaps/secrets/tokens still being kept updated seems correct.

Yeah, I definitely agree. I can't believe that no one noticed this sooner, but I guess most things stored as secrets have long enough expiration times that it was never an issue until short-lived kube service account tokens started getting used. (Or maybe this used to work in older versions?)

---

Since it seems like this is a complicated fix on the kubelet volume management side, for anyone else who encounters this issue we hacked around this by patching the apiserver to allow expired service account tokens from pods that are terminating, which isn't a terrible security compromise vs turning of expiration enforcement. This is the patch for 1.23 if anyone needs it: https://gist.github.com/PaulFurtado/2b8d8da2d8afe3ad375dd606b32a2297 should be easy to port to other versions too since it is so simple.

**Comment by DaGenix** [comment]
I believe this also an issue when running on AWS and attaching IAM roles to service accounts. In that case, AWS limits the token expiration to 1 day - meaning any Pod stuck in the terminating state for >1 day will run into this issue and I don't believe there is a workarond.

**Comment by liggitt** [comment]
I don't know kubelet volume management well enough to know how invasive this would be to fix, but the failure mode is pretty bad for a long preStop hook... are there any details about what would be required to change to resolve this?

**Comment by liggitt** [comment]
This got reported again in https://github.com/kubernetes/kubernetes/issues/122102

is there a plan to resolve the issue with volume plugins not refreshing during graceful termination?

**Comment by DaGenix** [comment]
@liggitt I attempted to resolve the problem in #117329 - however, there were some performance concerns noted. I've opened a new PR, #122161, that I hope resolves the performance concerns.

**Comment by alfredkrohmer** [comment]
FWIW we mitigated this problem by granting `create` on `serviceaccount/token` on and to the service account itself and "manually" requesting a new service account token with a lifetime slightly longer than the termination grace period via `kubectl create token` in the `preStop` hook script of the container. But obviously this is just a stopgap solution until this issue is properly fixed in code.

**Comment by gjkim42** [comment]
/cc 

This looks more important now as we introduced the native sidecar container(SidecarContainers feature) which should be running (and restarting) even during the pod termination until all the regular containers are stopped.

**Comment by drjackild** [comment]
I talked with Sergey and start looking at this issue

**Comment by DaGenix** [comment]
I have a potential fix for this issue: #122161. The challenge is that I need someone with the appropriate expertise to review the change to determine if its a reasonably solution, but I'm not sure how to contact the right folks.

**Comment by DaGenix** [comment]
FYI: ^ @SergeyKanzhelev @drjackild

**Comment by drjackild** [comment]
Hi @DaGenix! I'm looking on your fix and discussions, around it. That's great that you didn't abandoned it, I believe we find a proper way to finish it then, thank you!

**Comment by SergeyKanzhelev** [comment]
@drjackild would you please try to create an e2e test for this or perhaps a few e2e tests with the different scenarios? Reasoning on this part of logic is hard without proper test coverage

**Comment by drjackild** [comment]
> [@drjackild ](https://github.com/drjackild) would you please try to create an e2e test for this or perhaps a few e2e tests with the different scenarios? Reasoning on this part of logic is hard without proper test coverage

Sure, that's a great idea, will do 👍

**Comment by drjackild** [comment]
While I'm working on the e2e tests here is the minimum viable configuration to confirm bug:

https://gist.github.com/drjackild/e40d17de4e6e3f9b15679e8c2152034c

Steps to reproduce:

1. `kubectl apply -f token-bug-repro.yaml`
2. in a separate window open logs with `kubectl logs -f sa-token-test-pod`
3. `kubectl delete pod sa-token-test-pod`
4. after PreStop hook start, we should observe the correct behaviour where we're able to retrieve 200 OK response from the API
5. after sleep, hook should indicate, that the last request was failed because of the invalid token

**Comment by drjackild** [comment]
@SergeyKanzhelev FYI

I wrote E2E test to confirm that token is stopped rotating in the termination state.

https://github.com/kubernetes/kubernetes/compare/master...drjackild:kubernetes:sa-token-termination-rotation-e2e-test?expand=1

Test output: https://gist.github.com/drjackild/e2e759d1efef7a0eb5d2623a5a9afadf

**Comment by patrickvinograd** [comment]
Glad to see this is getting some renewed attention. We just ran into this in the context of using a projected service account token to do federated auth/token exchange from AWS to GCP. We have some long-running data processing jobs where we set a long terminationGracePeriod to allow the processing to complete, and were seeing sporadic failures when they attempt to auth to a GCP API. Narrowed it down to the projected token being expired, and this explains why.

**Comment by drjackild** [comment]
Today I confirmed, that my E2E test is passing on the @DaGenix branch, at least on the `./hack/local-up-cluster.sh` cluster.

@DaGenix does it makes sense to add e2e test directly to your branch? Your changes is actually much wider, in terms that they touch the whole volume mount logic for terminating pods, but I think it still make sense to ship them together, as they will fail on the master branch anyway

**Comment by DaGenix** [comment]
@drjackild That's awesome! I'll pull your test into my branch this evening

**Comment by DaGenix** [comment]
@drjackild - i added your commit to my branch: https://github.com/kubernetes/kubernetes/pull/122161. Some sort of CLA warning is now popping up.

**Comment by drjackild** [comment]
@DaGenix thanks! Looks like the CLA warning is about [your commit](https://github.com/kubernetes/kubernetes/commit/3f678956bf523bfc3618ef953e23c2a905112f21), did you signed CLA agreement?

**Comment by DaGenix** [comment]
@drjackild Oh no! It looks like when I rebased my commits it changed the author and I didn't realize it. I'm sorry I thought it was you. I've rebased again to fix the issue. Thanks for pointing it out!

**Comment by patrickvinograd** [comment]
@SergeyKanzhelev @natasha41575 Since this bug was triaged for SIG Node Bugs, and there's been a PR with a proposed fix (https://github.com/kubernetes/kubernetes/pull/122161) hanging out for several months with no feedback, is there any chance somebody in the SIG Node space can review/help push it forward?
